fix private repo creation in case of ignore templates & repo description bugs in case of template repos

g14a opened the pull request:
Fixes #3969
Fixes #3931

Followup to #3746

g14a committed:
fix private repo creation in case of ignore templates

g14a committed:
gofmt

mislav reviewed (commented):
@g14a Thanks for this! As an exercise for myself, I'm going to attempt to add some tests and tweaks to this to further ensure that there are no incompatibilities between REST and GraphQL repo creation. 👍

mislav committed:
Separate payload structs for REST vs GraphQL repo create

This enforces strict separation between serialization structs used for
repository creation payload with respect to whether GraphQL or REST was
used. Before, a field added to a GraphQL payload would leak to REST
payload (and vice versa).

g14a committed:
Merge branch 'trunk' of https://github.com/cli/cli into fix/private-repo-create

g14a committed:
fix description related bugs in creating a template repo

g14a commented:
@mislav I tried fixing #3931 as well and pushed a commit because it is just a one line fix(and because it is similar to #3969). Hope its not a problem. Make sure to pull the latest commit for your tweaks :)

mislav committed:
Merge remote-tracking branch 'origin' into fix/private-repo-create

mislav commented:
I've pushed a refactoring that creates strictly separate payload structs for REST vs GraphQL repository creation. Before, the same `repoCreateInput` struct was used for JSON serialization to both REST and GraphQL endpoints, which led to invalid JSON parameter names being sent in REST mode.

Summary of changes:

* Fixes REST payload parameter names when gitignore or license are used
* Fixes setting repository visibility when gitignore/license were used for org repos
* Fixes setting team ID when gitignore/license were used for org repos
* Fixes assertions in tests for create flow with gitignore and license
* Adds extensive `repoCreate()` unit tests

Please review!

samcoe commented on pkg/cmd/repo/create/create.go:
```diff
@@ -257,15 +257,27 @@ func createRun(opts *CreateOptions) error {
+		var templateRepo ghrepo.Interface
 		apiClient := api.NewClientFromHTTP(httpClient)
 
 		cloneURL := opts.Template
```
I know that this was already in place but I think it would be clearer to rename this to `templateRepoURL` or something similar since `toClone` was renamed to `remplateRepo`.

samcoe commented on pkg/cmd/repo/create/http.go:
```diff
@@ -10,79 +10,102 @@ import (
-	Name       string `json:"name"`
-	Visibility string `json:"visibility"`
-	OwnerID    string `json:"ownerId,omitempty"`
+type createRepositoryInput struct {
```
It isn't immediately obvious from the name what the difference between `repoCreateInput`, `createRepositoryInputV3`, and `createRepositoryInput` is, I think comments about their purpose could help clear that up.

samcoe commented on pkg/cmd/repo/create/http.go:
```diff
@@ -104,34 +127,55 @@ func repoCreate(client *http.Client, hostname string, input repoCreateInput, tem
+		path := "user/repos"
+		if isOrg {
+			path = fmt.Sprintf("orgs/%s/repos", input.OwnerLogin)
+			inputv3.Visibility = strings.ToLower(input.Visibility)
```
This is not a CLI issue but the existence of `private` and `visibility` in this API endpoint is confusing.

samcoe commented on pkg/cmd/repo/create/http_test.go:
```diff
@@ -1,37 +1,345 @@
-	input := repoCreateInput{
-		Description: "roasted chestnuts",
-		HomepageURL: "http://example.com",
+func Test_repoCreate(t *testing.T) {
```
These tests are great, just for completeness I think one case that is missing is creating a repo with a hostname that is not `github.com`.

samcoe commented on pkg/httpmock/stub.go:
```diff
@@ -100,6 +100,17 @@ func FileResponse(filename string) Responder {
 	}
 }
 
+func RESTPayload(responseStatus int, responseBody string, cb func(payload map[string]interface{})) Responder {
```
Yes! 👍

samcoe reviewed (approved):
Left a couple non-blocking suggestions. This looks great, much clearer when separating out the GraphQL and REST payloads.

mislav committed:
:nail_care: address review comments
